core/state_transition.go

```diff
@@ -253,7 +253,13 @@ func (st *StateTransition) buyGas() error {
 
 func (st *StateTransition) preCheck() error {
 	if st.msg.IsDepositTx() {
-		// No fee fields to check, no nonce to check, and no need to check if EOA (L1 already verified it for us)
+		// No fee fields to check, and no need to check if EOA (L1 already verified it for us)
+		if st.msg.Nonce() != 0 {
+			// DepositTx always has a 0 nonce, but when deserializing JSON it may be a wrapper that holds the "real"
+			// nonce that was actually used. We must ensure nonce is 0 for any tx that are being exeucted
+			return fmt.Errorf("%w: address %v, tx: %d", ErrNonceTooHigh,
+				st.msg.From(), st.msg.Nonce())
+		}
```

Comment:
Do we need this check if we decide we don't introduce `depositTxWithNonce`? Then the deposit-tx nonce is always zero  https://github.com/ethereum-optimism/op-geth/blob/a4026243c05728b288367eb165b46bb9a6f9248f/core/types/deposit_tx.go#L77

Reply:
Took a while, but I see the product problem in the `types.Transaction` go usage now. So nvm this comment.